src/Lykke.AlgoStore.CSharp.AlgoTemplate.Services/Services/UserLogService.cs

```diff
@@ -11,28 +12,92 @@ namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Services.Services
     /// <summary>
     /// <see cref="IUserLogService"/> implementation
     /// </summary>
-    public class UserLogService : IUserLogService
+    public class UserLogService : IUserLogService, IDisposable
     {
-        private readonly ILoggingClient _userLogRepository;
+        [NotNull] private readonly ILoggingClient _userLogClient;
+        [NotNull] private readonly BatchBlock<UserLogRequest> _batchBlock;
+        [NotNull] private readonly ActionBlock<UserLogRequest[]> _actionBlock;
 
-        public UserLogService(ILoggingClient userLogRepository)
+        private readonly TimeSpan _maxBatchLifetime;
+        private DateTime _currentBatchExpirationMoment;
+        private readonly Timer _timer;
+
+        public UserLogService([NotNull] ILoggingClient userLogClient,
+            TimeSpan maxBatchLifetime,
+            int batchSizeThreshold)
+        {
+            if (maxBatchLifetime < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxBatchLifetime), maxBatchLifetime,
+                    "Should be positive time span");
+            }
+
+            if (batchSizeThreshold < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(batchSizeThreshold), batchSizeThreshold,
+                    "Should be positive number");
+            }
+
+            _maxBatchLifetime = maxBatchLifetime;
+            _userLogClient = userLogClient;
+
+            _batchBlock = new BatchBlock<UserLogRequest>(batchSizeThreshold);
+
+            _actionBlock = new ActionBlock<UserLogRequest[]>(x => PersistUserLogs(x),
+                new ExecutionDataflowBlockOptions {MaxDegreeOfParallelism = 2});
+
+            _batchBlock.LinkTo(_actionBlock);
+
+            ExtendBatchExpiration();
+
+            //Initiate timer after 1s (1000ms) with a period of 10s (10000ms)
+            _timer = new Timer(HandleTimerTriggered, null, 1000, 10000);
```

Comment:
Why do we need this timer?

Reply:
Timer is used to make sure that logs will be saved in timely fashion even when log batch is not full

Reply:
What happens if the max batch lifetime is shorter than 10s?

Reply:
I've changed timer period to 50ms